# self_learning_bulk_rna-seq
# Notes and sources

## Best practices when coding with R
https://r4ds.had.co.nz/

## Background and R coding
Want to start off with this resource as a basis for making a pipeline

https://doi.org/10.12688/f1000research.7035.2



## FASTQC
https://rtsf.natsci.msu.edu/genomics/tech-notes/fastqc-tutorial-and-faq/



## Different ways to do GSEA, manual, python, and R

How to input data for overall cell state (need normalized counts matrix)
https://www.youtube.com/watch?v=KY6SS4vRchY&t=2004s	

preranked protocol
https://docs.google.com/document/d/17cp4YuFwy80NLrlPTSRePnF2dePpUmV3178NGE5s4rk/edit

GSEA in python	n	preranked, easy to use, unsure if this is best
https://www.youtube.com/watch?v=yOQcrUMCALw	

GSEA in R (clusterProfiler),	preranked, unsure if this is best
https://learn.gencore.bio.nyu.edu/rna-seq-analysis/gene-set-enrichment-analysis/	

## Sources to sift through
https://cumulus.readthedocs.io/en/stable/bulk_rna_seq.html
https://www.encodeproject.org/data-standards/rna-seq/long-rnas/
https://github.com/ENCODE-DCC/rna-seq-pipeline
https://github.com/ENCODE-DCC/rna-seq-pipeline/blob/dev/docs/reference.md
https://bioinformatics-core-shared-training.github.io/RNAseq-R/align-and-count.nb.html
https://bioconductor.org/packages/release/bioc/vignettes/Rsubread/inst/doc/SubreadUsersGuide.pdf
https://bioinformatics-core-shared-training.github.io/RNAseq-R/align-and-count.nb.html
https://www.bioconductor.org/packages/devel/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#experimental-data
ivg.org
https://scienceparkstudygroup.github.io/rna-seq-lesson/04-bioinformatic-workflow/index.html#24-counting-and-sorting
https://rtsf.natsci.msu.edu/genomics/tech-notes/fastqc-tutorial-and-faq/
